[WIP] Sparse PCA online in the direction of samples

arthurmensch opened the pull request:
Following #4856, I implemented a dictionary learning algorithm that output a dense code and a sparse dictionary, which corresponds to the description of SPCA given in [1]. It alternated between ridge regression for learning code for a batch of samples, and projected block coordinate descent on the elastic net ball.

For each features (i.e. pixel in faces example) we project the `n_component` vector that contains all dictionary component value for this feature, which explicitly enforce dictionary components to be non overlapping (this differ from [1], where projection occurs in the feature space and not the component space)

For the moment there is no unit tests, but output of the method can be seen running `example.decomposition.lot_decomposition_faces`. As we use a large outer loop of `n_feature` iterations and an inner loop for elastic net projection, I wrote a cython extension for speed consideration.

Does this method seem to have an interest for the project ?

Mairal, J., Bach, F., Ponce, J., & Sapiro, G. (2010). Online learning for matrix factorization and sparse coding. The Journal of Machine Learning Research, 11, 19-60.

arthurmensch committed:
Elastic net constraint on dictionary learning for new online sparse PCA

ogrisel commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -56,6 +57,7 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
         omp: uses orthogonal matching pursuit to estimate the sparse solution
         threshold: squashes to zero all coefficients less than regularization
         from the projection dictionary * data'
+        linear: uses a non-penalized least square fit (regularization parameter is ignored)
```
`ols` or `least_squares` would be a better (more explicit) name in my opinion.

ogrisel replied:
Actually this will not yield a sparse code, so I wonder if adding this option here does not reflect a design problem.

ogrisel commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -132,9 +134,17 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
+    elif algorithm == 'linear':
+        # Should be put into sparse_encode
+        # alpha = float(regularization) / n_features
+        lr = Ridge(alpha=0.0001 / n_features, fit_intercept=False, normalize=False)
```
Why not use `LinearRegression`?  Is the ridge penalty beneficial in terms of speed or stability of the model? It seems to contradict the docstring.

arthurmensch replied:
I just put it for security but it appears to be useless with faces dataset

ogrisel commented on sklearn/utils/enet_proj.py:
```diff
@@ -0,0 +1,59 @@
+__author__ = 'arthur'
```
Please follow the conventions in other headers of files of the project to state authorship consistently.

ogrisel commented on sklearn/utils/enet_proj.py:
```diff
@@ -0,0 +1,59 @@
+    return np.sum(b_abs * (l1_ratio + (1 - l1_ratio) * b_abs))
+
+
+def enet_proj(b, tau=1, l1_ratio=0.1, random_state=None):
```
This code is both interesting an non trivial. Please add a reference to the paper that best describe it. Also please add unit tests dedicated to check the correctness of this implementation on a bunch of (deterministically seeded) random vectors with a elastic net radius scaled to be large enough to lie outside of the ball defined by `l1_ratio` and `tau`. 

Actually one should find a more explicit name for the `tau` variable. For instance `radius`.

Also please use explicit function names such as `enet_projection` and finally make those functions private (`_enet_projection`).

ogrisel commented on examples/decomposition/plot_faces_decomposition.py:
```diff
@@ -99,10 +105,18 @@ def plot_gallery(title, images, n_col=n_col, n_row=n_row):
      True),
+
+    ('Elastic Net Sparse PCA',
+     decomposition.MiniBatchDictionaryLearning(n_components=n_components, alpha=alpha,
```
The name of the method and the class do not match.

ogrisel commented:
I think this method is interesting but I think the class organisation needs a rework:

- the current MiniBatchSparsePCA does minibatches in the feature space, I think this is an abuse. At least it's not consistent with other use of the `MiniBatch` prefix in scikit-learn. I would rather have called this BlockedCoordinateSparsePCA or something similar.

- the new method would implement Sparse PCA with mini-batches in the sample space (via ENet projections of the components after each minibatch update if I understand correctly) but is currently implemented as part of the online dictionary learning code base for code reuse purpose.

However sparse PCA is not dictionary learning as the activations / code are not sparse. From a user point of you I find this really confusing. I think we should find a way to refactor the code to have the estimator name reflect the model name (the class of objective function with either sparse components or sparse code) and not the kind of solver used internally.

ogrisel commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -1036,18 +1108,23 @@ class MiniBatchDictionaryLearning(BaseEstimator, SparseCodingMixin):
         total number of iterations to perform
 
-    fit_algorithm : {'lars', 'cd'}
+    fit_algorithm : {'lars', 'cd', 'elastic_net'}
```
I think this should be called `projected_gradient`. elastic_net by itself is not a solver like lars and coordinate descent but rather a kind of penalty.

ogrisel replied:
Also does is the current implementation well defined and equivalent to l1-ball projection when `l1_ratio=1.0`?

arthurmensch replied:
It should. Do we have existent l1-ball projection to compare with ?

arthurmensch replied:
It does

ogrisel commented:
Out of curiosity what is the speed impact of using the python version of the Elastic Net projection? If the use of the the cython version is critical for speed, I think we should just write tests for that version and completely drop the Python version from the code base. There is no point in keeping 2 versions if one is always sub optimal.

arthurmensch commented:
We gain a factor ~40 in performance on the face dataset. I kept the python version in the database for testing, but it should be included in a test file.

arthurmensch commented:
Actually the online SPCA method presented in this PR can also enforce code sparsity : however it is often the case that you do not want both sparse code and sparse dictionary.

I think we could refactor as follow :
- Allow user selection of code computation algorithm, ie `{'cd', 'lars', 'ols'}`, such allowing him to use a non sparse code. This would involve a parameter `alpha` controlling code sparsity.
- Allow user selection of dictionary constraint, i.e. `{'ordinary', 'sparse'}`, with a parameter `l1_ratio` controlling dictionary sparsity.

`MiniBatchDictionaryLearning` would then default to sparse code, dense dictionary, while MiniBatchSparsePCA would default to sparse dictionary, sparse code, which is its expected behaviour as for version 0.16.

We could keep a class MiniBatchFeatureSparsePCA (or what not...) that would keep computing a sparse code and a dense dictionary by transposing `X`.

I still need to do some benchmarking to see if we do not loose considerable performance depending on `X` shape.

One of the problems with this approach is that the output of `fit` using new `MiniBatchDictionaryLearning` would be different using the same parameter, as there is no obvious correspondence between regulation parameters `l1_ratio` and `alpha` (which is used today to regulate code sparsity and thus `component_` sparsity in MiniBatchSparsePCA).

arthurmensch commented:
Sorry for the multiple commits, to sum up I added a tolerance based early stopping, using the surrogate function $\hat f (t)$ described in above reference, and I changed the API so we can control code learning algorithm ('lars, 'cd' (not working on master cf previous PR), 'ols'), and dictionary learning constraints (either euclidian or enet, which is structure inducing).

arthurmensch commented:
I tweaked the cython codeso that it can release the gil. The for loop in `_update_dict_enet` is embarissingly parallel, are we allowed to use `prange` or should we always stick to joblib ?

ogrisel commented:
`prange` cannot be currently used in scikit-learn because:
- not all supported compilers (e.g. clang on OSX or MSVC 2008 on windows) implement OpenMP
- the OpenMP implementation of gcc (libgomp) is still crashing when used in a program that uses multiprocessing somewhere else: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60035

However you can either use joblib with `backend="threading"` if you do not do many calls to parallel in a row (otherwise the thread pool creation and destruction will be killing the performance). Alternatively you can maintain you own pool (the use of contextlib is necessary to use the ThreadPool as a context manager in Python 2.6):

```python
from contextlib import closing
from multiprocessing.pool import ThreadPool

with closing(ThreadPool(n_jobs)) as p:
    for X_batch in minibatch_iterator:
        # use the pool to apply a function in parallel for instance p.map

```

ogrisel commented:
Note that we could also work to make `joblib.Parallel` a context manager with `__enter__` and `__exit__` methods to keep a managed pool internally for successive jobs. This requires a lot more work and would not bring any benefit fro the threading backend case. It could be useful for the multiprocessing backend though.

arthurmensch commented:
Since we are only using a ThreadPool for `enet` version, I explicitly close at the end of `dict_learning_online`. I also tried to use the constructed Pool in `sparse_encode`, without gain in performance (I guess we need a multiprocessing Pool for that).

ogrisel commented:
I changed this PR to `[WIP]` as we need to resolve the API issue. I suggest we introduce classes with name `IncrementalDictionaryLearning` and `IncrementalSparsePCA` to host the implementation of the cases where we have a `partial_fit` method and keep `DictionaryLearning` and `SparsePCA` for the cases where we cannot implement a `partial_fit` method (even if the `fit` method does some kind of internal 'online' optimization in the axis of the features).

We could then deprecate the current `MiniBatchDictionaryLearning` and `MiniBatchSparsePCA` as they are introducing confusion as they are not consistent.

ogrisel commented:
@arthurmensch when you rebase and squash commit, please cleanup the full commit message: they are old phony messages included in the commit message for 0fbc54b.

Commit:
test

arthurmensch commented:
The major issue here is now a naming problem : this PR code could be used to change MiniBatchSparsePCA into a real online estimator.

However, we would have to keep present MiniBatchSparsePCA for the next few versions and deprecate it, while creating another estimator (IncrementalSparsePCA ?)

arthurmensch commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -94,7 +101,7 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
 
     if algorithm == 'lasso_lars':
-        alpha = float(regularization) / n_features  # account for scaling
+        alpha = float(regularization) / sqrt(n_features)  # account for scaling
```
This scaling is the one recommended in the Lasso litterature. I can write a test showing that sparsity elvel stays roughly the same no matter what input feature size is provided.

arthurmensch commented:
Output from example :
```
Dataset consists of 400 faces
Extracting the top 30 Online Dictionary learning...
Iteration   0 (elapsed time:   0s,  0.0mn)
Iteration  10 (elapsed time:   0s,  0.0mn)
Iteration  20 (elapsed time:   0s,  0.0mn)
Iteration  30 (elapsed time:   0s,  0.0mn)
Iteration  40 (elapsed time:   0s,  0.0mn)
Iteration  50 (elapsed time:   1s,  0.0mn)
Iteration  60 (elapsed time:   1s,  0.0mn)
Iteration  70 (elapsed time:   1s,  0.0mn)
Iteration  80 (elapsed time:   1s,  0.0mn)
Iteration  90 (elapsed time:   2s,  0.0mn)
done in 2.416s
Online Dictionary learning - Component density
0.0601806640625
Code density
1.0
```

![components](https://cloud.githubusercontent.com/assets/4078691/8647593/a258cc42-2959-11e5-8a06-807b3c0ee556.png)
![reconstruction](https://cloud.githubusercontent.com/assets/4078691/8647592/a25842d6-2959-11e5-8471-7a6f0149534d.png)

Commit:
Merge branch 'master' of github.com:scikit-learn/scikit-learn

Commit:
Normalizing dictionary before sparse coding

Commit:
Normalizing dictionary before sparse coding

Commit:
Normalizing dictionary before sparse coding

Commit:
WIP

Commit:
Tweak for initial projection

Commit:
WIP

Commit:
WIP

arthurmensch commented:
I am closing this PR for the moment as it needs serious rework from my side

amueller commented:
ok thanks for working on this :)
